include/engine/routing_algorithms/routing_base.hpp

```diff
@@ -524,7 +528,11 @@ template <class DataFacadeT, class Derived> class BasicRoutingInterface
                                           std::vector<NodeID> &packed_path) const
     {
         NodeID current_node_id = middle_node_id;
-        while (current_node_id != search_heap.GetData(current_node_id).parent)
+        // all initial nodes will have itself as parent
+        // all incomplete paths (e.g. paths ending at the core boundary) will have SPECIAL_NODEID
+        // as parent
+        while (current_node_id != search_heap.GetData(current_node_id).parent &&
+               search_heap.WasInserted(search_heap.GetData(current_node_id).parent))
```

Comment:
The code location that actually breaks is the loop code: https://github.com/Project-OSRM/osrm-backend/blob/e91bfebf237c3adfa26b8077377c54a5e3dd9d54/include/engine/routing_algorithms/routing_base.hpp#L93-L94

If you mark the core nodes like we do with real start nodes (node.parent == node), we trigger the loop logic for them as well. If for a core node this finds a shorter path that we never wanted to search for,  this will lead to a crash in the path unpacking as it will return a path that is (middle, middle)  but middle is not the forward for backward node of the source phantom.

Hence we need to have a way to differentiate between real start nodes (for which we want to allow loops) and for core start nodes (for which we don't want loops). 

I can dropbox you the data file and breaking query as it is proprietary data.